### DISCUSSION TITLE
Cannot find library that is on the system

### DISCUSSION THREAD

**Discussion by myz-dev** [rust-cross/cargo-zigbuild#137] [Q&A] [answered]
Hi,

`using WSL Ubuntu 20.04.6 LTS`

for a while now I have been trying to cross compile my project with a specific GLIBC version on my target system.
I think `zigbuild` is exactly what I am looking for and I tried to follow the instructions and compile my program.

From the root directory of my project I run 
`cargo zigbuild --bin defect_map --release --target i686-unknown-linux-gnu` (omitting the specific GLIBC version for the minimal example) and the compilation fails at the very last step, requiring the SQLITE development library: `= note: ld.lld: error: unable to find library -lsqlite3`.

I have checked my installation of the package `dpkg -L libsqlite3-dev:i386` and everything is where it is supposed to be.

The library for the target system is under `/usr/lib/i386-linux-gnu` which is also configured in `/etc/ld.so.conf.d/i386-linux-gnu.conf`. Similarly the library exists for my host system, but I do not know if this is relevant for the cross compilation step.

So how can I tell `zigbuild` where to look for the library?

Best regards

**Comment by NobodyXu** [comment]
It's likely you are missing pkg-config for this

**Reply by myz-dev** [reply]
Hi,
thank you for your swift reply.
I have the correct pkg-config file on my system. It was automatically installed to `/usr/lib/i386-linux-gnu/pkgconfig/`.
I checked the contents and it seems OK.

I am trying to compile for the `i686-unknown-linux-gnu` target. 
I am calling zigbuild like this:
`PKG_CONFIG_PATH="/usr/lib/i386-linux-gnu/pkgconfig/:${PKG_CONFIG_PATH}" cargo zigbuild --bin my_binary --release --target i686-unknown-linux-gnu`

I still get the error, that my library could not be found. Do you have any more hints on why that could be? Could there be any confusion between the `i386` of the available architecture and the `i686` of the available `rustc` target?
Thank you again for your first hint.

**Reply by NobodyXu** [reply]
Can you try running pkg-config directly to see if it actually works?

**Reply by myz-dev** [reply]
Hi,
running `pkg-config --version` prints out `0.29.1`.
And running `PKG_CONFIG_PATH="/usr/lib/i386-linux-gnu/pkgconfig/:${PKG_CONFIG_PATH}" pkg-config libsql:i386` exits successfully without printing anything (it exits with an error code if I run it without the path, which I think is expected, as per default it only looks at `/usr/lib/pkconfig`).

**Reply by NobodyXu** [reply]
Can you try if pkg-config can give you correct information for the libsqlite3 installed, by passing "--cflags --libs" to pkg-config?

**Reply by myz-dev** [reply]
Hi, thank you again!
I have tested this:
`PKG_CONFIG_PATH="/usr/lib/i386-linux-gnu/pkgconfig/" pkg-config libsqlite3 --cflags --libs` and got the error message, that `libsqlite3.pc` was not found. The `pc` file installed by `apt` (using Ubunut 20.04) was called `sqlite3.pc`. So I copied that file to the name that was needed and the command now prints `-lsqlite3` instead of the error message.

Trying to build with 
`PKG_CONFIG_PATH="/usr/lib/i386-linux-gnu/pkgconfig/:${PKG_CONFIG_PATH}" cargo zigbuild --bin my_binary --release --target i686-unknown-linux-gnu`
still gives me the same error:
`= note: ld.lld: error: unable to find library -lsqlite3`

**Reply by NobodyXu** [reply]
I think the `libsqlite3.pc` is problematic, it does not specify `-L /usr/lib/i386-linux-gnu/`, which I presume is where the `libsqlite3.{so, a}` is installed and is not included in the default library search path of compiler/linker.

**Reply by myz-dev** [reply]
Hi.
My file `/usr/lib/i386-linux-gnu/pkgconfig/libsqlite3.pc` looks like this:
```
 # Package Information for pkg-config

prefix=/usr
exec_prefix=${prefix}
libdir=${prefix}/lib/i386-linux-gnu
includedir=${prefix}/include

Name: SQLite
Description: SQL database engine
Version: 3.31.1
Libs: -L${libdir} -lsqlite3
Libs.private: -lm -lm -ldl -lz -lpthread
Cflags: -I${includedir}

```
I think `libdir` resolves to `/usr/lib/i386-linux-gnu` where the library is installed. Do you think I have to change this line?

**Reply by NobodyXu** [reply]
I suspect that `pkg-config` loads the wrong `libsqlite3.pc`.

You can try setting `PKG_CONFIG_PATH="/usr/lib/i386-linux-gnu/pkgconfig/"`, it might help.

**Reply by myz-dev** [reply]
Unfortunately that does not help. The library is not found during the build process. Thank you very much for your kind help so far, I will see what I can do about this :-)

**Reply by NobodyXu** [reply]
You can also try setting `PKG_CONFIG` to your own script that invokes `pkg-config` with the correct parameter.

**Reply by myz-dev** [reply]
Ok, thank you, I will check how that works and report the outcome here. Thank you again for your time!

**Comment by polarathene** [comment, marked as answer]
> The library for the target system is under `/usr/lib/i386-linux-gnu` which is also configured in `/etc/ld.so.conf.d/i386-linux-gnu.conf`.

`/etc/ld.so.conf.d/*` is for building `/etc/ld.so.cache` DB, if you add a new config there you would need to run `ldconfig` to update the cache. However this is only for adding search paths for runtime linking (_when a program starts, or you use `ldd` for example_), similar to how `LD_LIBRARY_PATH` is for runtime not build time.

---

> `= note: ld.lld: error: unable to find library -lsqlite3`

This is a link time error during build.

If you do not experience this problem with `cargo build` (_perhaps try with matching host arch to verify compared to `cargo zigbuild` for that same host arch_), then you likely just need to ensure you provide the expected search dir.

Are you certain the issue is with `pkg-config`? If it was at link time with Zig instead, when `zig cc` is provided a `-target` it will disable using system default search paths IIRC, so you need to add `RUSTFLAGS='-L /usr/lib/i386-linux-gnu'` if that is where your `libsqlite3.so` / `libsqlite3.a` exists (_which the error is trying to find for linking_).

**Reply by myz-dev** [reply]
Hi, thank you very much for your reply!

A lot of time has passed since and unfortunately I do not have access to the project anymore.

As far as I remember, I had tried to solve this with `cross` and/or `zigbuild` for quite some time and, at the time, just was not able to. I am a little ashamed to say, that, if I remember correctly, to compile my project for the specific target, I created a container using images that had old tool chains installed, that targeted the specific glib-c version I needed.

**Reply by polarathene** [reply]
No shame, this type of task can get complicated and not always documented well. I am troubleshooting one situation myself where a `-sys` crate is building with zig correctly, but when another crate is built with that `-sys` crate as a dependency there is something else preventing it from respecting the older glibc version I configured zig target to use.

Sometimes it easier and more practical to just use what works, even if it is not ideal :)
